src/svgp.jl

```diff
@@ -77,6 +77,72 @@ function StatsBase.mean_and_var(f::ApproxPosteriorGP{<:SVGP}, x::AbstractVector)
     return μ, Σ_diag
 end
 
+raw"""
+    pathwise_sample(
+        [rng::AbstractRNG,]
+        f::ApproxPosteriorGP{<:SVGP},
+        x::AbstractVector,
+        prior_sample_function,
+        num_samples=1::Int
+    )
+
+Takes a 'pathwise sample' from the posterior GP `f` at `x` by using Matheron's
+rule [2]. This works by taking `num_samples` (possibly approximate) samples from
+`f.prior` and then updating these prior samples with samples `u` from
+`f.approx.q` by Matheron's rule:
+
+```math
+f^* | u = f^* + K_{*, u} K_{u, u}^{-1} (u - f^z)
+```
+
+where ``f^* = f(x^*)`` and ``f^z = f(z)`` are the prior samples at the test
+points and inducing points respectively. ``u \sim q(u)`` is a sample from the
+variational posterior. ``K_{u, u}`` and ``K_{*, u}`` are the prior covariances
+for the inducing points and for between the test and inducing points
+respectively.
+
+`prior_sample_function` can be any function which returns exact or approximate
+samples from the prior. It must have the signature:
+
+`prior_sample_function(rng::AbstractRNG, fx::FiniteGP, num_samples::Int)`
+
+[2] - James T. Wilson and Viacheslav Borovitskiy and Alexander Terenin and Peter
+Mostowsky and Marc Peter Deisenroth. "Efficiently Sampling Functions from
+Gaussian Process Posteriors" ICML, 2020.
+"""
+function pathwise_sample(
```

Comment:
In terms of design, it seems a shame to require that the locations at which the sample it to be evaluated are needed when the sample is drawn. I wonder whether it would be doable to make this function return a function that can be evaluated at any new input / collection of inputs? So the signature would become something like
```julia
struct ApproxPostSample
    
end

(f::ApproxPostSample)(x) = prior(x) + update(x)

# Optimisation for multiple prediction locations.
map(f::ApproxPostsample, x::AbstractVector) = map(prior, x) + map(update, x)

function pathwise_sample(
    rng::AbstractRNG,
    f::AbstractPosteriorGP{<:SVGP},
    prior_sample_function;
    num_samples=1::Int,
)
    ...
    return ApproxPostSample(prior_sample, terms_needed_for_posterior_update...)
end
```

Reply:
That's more or less the original version I had - correct me if I'm wrong but isn't sampling functions like that only possible if you use some kind of a weight space approximation? (i.e you wouldn't be able to do it if you were just using `rand` right?) So, I wrote it this way to allow you to use literally any kind of prior sample function.

I agree though - perhaps that's too general to the point of uselessness. It might then be better to have this function in the RFF package/whatever instead?

Reply:
So I guess what I'm saying is that, in my opinion, your original approach is something that would be cool to have in this package, even if we also have to define an RFF thing.

Reply:
> I worry that might be the case -- it's unclear to me that, under the current API, much is added over just writing rand(f(x)) or whatever. Or is this not an accurate reflection of the benefits?

I think it's actually strictly worse if you don't use an approximation :smile: - you're right, there's not much point in doing it this way, I'll change it to a function sample version.